File: app/src/main/java/org/alkaline/taskbrain/data/UserDocSignal.kt
package org.alkaline.taskbrain.data

import android.util.Log
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

/**
 * Per-user signal observed by `NoteStore`'s listener to trigger delta pulls.
 * Fire-and-forget — never awaited by callers. See `docs/live-cross-platform-sync.md`.
 */
object UserDocSignal {
    private const val TAG = "UserDocSignal"

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** Called after every successful notes-collection write. */
    fun bump(db: FirebaseFirestore, userId: String): Job =
        setUserField(db, userId, "lastNoteChange" to FieldValue.serverTimestamp(), "UserDocSignal.bump")

    /** Create-on-login so the signal listener has a doc to attach to. */
    fun ensureExists(db: FirebaseFirestore, userId: String): Job =
        setUserField(db, userId, "uid" to userId, "UserDocSignal.ensureExists")

    private fun setUserField(
        db: FirebaseFirestore,
        userId: String,
        field: Pair<String, Any>,
        operation: String,
    ): Job = scope.launch {
        try {
            db.collection("users").document(userId)
                .set(mapOf(field), SetOptions.merge())
                .await()
            FirestoreUsage.recordWrite(operation, FirestoreUsage.WriteType.SET)
        } catch (e: Exception) {
            Log.w(TAG, "$operation(userId=$userId) failed", e)
        }
    }
}

File: web/src/data/UserDocSignal.ts
import {
  doc,
  serverTimestamp,
  setDoc,
  type Firestore,
} from 'firebase/firestore'
import { firestoreUsage } from './FirestoreUsage'

/**
 * Per-subsystem signal fields on `users/{uid}` that subsystem listeners
 * watch to trigger delta pulls. Each [Channel] owns its own timestamp field
 * on the same user doc; listeners observe their channel's field and ignore
 * others. Fire-and-forget — callers should not await.
 * See `docs/live-cross-platform-sync.md`.
 */
export type Channel = 'NOTES' | 'ALARMS'

const CHANNEL_FIELD: Record<Channel, string> = {
  NOTES: 'lastNoteChange',
  ALARMS: 'lastAlarmChange',
}

/** Field name used on `users/{uid}` for a given signal channel. */
export function channelFieldName(channel: Channel): string {
  return CHANNEL_FIELD[channel]
}

export const UserDocSignal = {
  /** Called after every successful write to the channel's collection. */
  bump(
    db: Firestore,
    userId: string,
    channels: Channel | Iterable<Channel>,
  ): Promise<void> {
    const list = typeof channels === 'string' ? [channels] : Array.from(channels)
    if (list.length === 0) return Promise.resolve()
    const fields: Record<string, unknown> = {}
    for (const c of list) fields[CHANNEL_FIELD[c]] = serverTimestamp()
    return setUserFields(db, userId, fields, `UserDocSignal.bump.${list.join('+')}`)
  },

  /** Create-on-login so signal listeners have a doc to attach to. */
  ensureExists(db: Firestore, userId: string): Promise<void> {
    return setUserFields(db, userId, { uid: userId }, 'UserDocSignal.ensureExists')
  },
}

async function setUserFields(
  db: Firestore,
  userId: string,
  fields: Record<string, unknown>,
  operation: string,
): Promise<void> {
  try {
    await setDoc(doc(db, 'users', userId), fields, { merge: true })
    firestoreUsage.recordWrite(operation, 'SET')
  } catch (e) {
    console.warn(`[UserDocSignal] ${operation}(userId=${userId}) failed`, e)
  }
}
